[ML-7905] Move image pull secret enrichment to server side

adamdelman opened the pull request:
[ML-7905] Move image pull secret enrichment to server side

Disable setting the image pull secret on the client side, as this can change during runtime. Instead, do it in the ServerSideLauncher so we always get the up to date parameter.
Move the test `test_new_function_args_with_default_image_pull_secret` to `services.api.tests.unit.test_launcher` as enrichment as done in the server side launcher.
Remove the test `services.api.tests.unit.runtimes.test_spark.TestSpark3Runtime.test_deploy_with_image_pull_secret` as enrichment is no longer done on the client side.
Add `__deepcopy__` method to the `Config` class to prevent looping when accessing the `_missing` field.

https://iguazio.atlassian.net/browse/ML-7905

[ML-7905]: https://iguazio.atlassian.net/browse/ML-7905?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into development

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into development

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into development

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into development

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into development

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into development

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into development

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into development

adamdelman committed:
[ML-7905] Move image pull secret enrichment to server side

Disable setting the image pull secret on the client side, as this can change during runtime. Instead, do it in the ServerSideLauncher so we always get the up to date parameter.

adamdelman committed:
fix

liranbg commented on server/py/services/api/launcher.py:
```diff
@@ -239,6 +240,15 @@ def _pre_run_node_selector_enrichement(self, runtime, run):
+        This ensures the image pull secret is correctly set on the run
+        while maintaining the runtime's integrity from system-specific project settings.
+        """
+        run.spec.image_pull_secret = mlrun.config.config.function.spec.image_pull_secret
```
```suggestion
        run.spec.image_pull_secret = mlrun.config.config.function.spec.image_pull_secret
```

the side effect is that if client set an explicit image_pull_secret you override it

adamdelman replied:
fixed

adamdelman committed:
fix

adamdelman committed:
fix

adamdelman committed:
fix test

adamdelman committed:
fix test

liranbg commented on tests/runtimes/test_run.py:
```diff
@@ -262,13 +262,12 @@ def test_with_limits():
 
 
 def test_new_function_args_with_default_image_pull_secret():
```
the test isnt accurate anymore 
before: it verified the default mlconf kicked in
after: it verifies client override works

I think what needs to be done here is moving this test to when function is created at api side, it is enriched with the configuration

adamdelman replied:
fixed

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into adamd/ML-7905

adamdelman committed:
fix test

adamdelman committed:
fix test

adamdelman committed:
fix test

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into adamd/ML-7905

adamdelman committed:
fix

liranbg commented on server/py/services/api/tests/unit/assets/sample_function.py:
```diff
@@ -0,0 +1,21 @@
+#
+import mlrun
+
+mlrun.mlconf.feature_store.flush_interval = None
```
you dont need that

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into adamd/ML-7905

liranbg reviewed (commented):
minor

adamdelman committed:
fix

adamdelman committed:
fix

adamdelman committed:
Merge branch 'development' of github.com:mlrun/mlrun into adamd/ML-7905

adamdelman committed:
Fix infinite recursion when deepcopying Config

When copy.deepcopy was called, Python tried to recursively access nested
Config objects via __getattr__, resulting in an infinite loop. This commit
adds a custom __deepcopy__ method to bypass the problematic lookup,
avoiding endless recursion.

adamdelman committed:
Remove irrelevant test

adamdelman committed:
fix
